src/types/timestamp.rs

```diff
@@ -1007,230 +973,212 @@ impl TimestampBuilder {
+        self.fractional_seconds = Some(Mantissa::Digits(3));
+        self.nanoseconds = Some(millisecond * 1_000_000);
+        self.change_state()
     }
 
     pub fn with_nanoseconds_and_precision(
-        self,
+        mut self,
         nanoseconds: u32,
         precision_digits: u32,
-    ) -> FractionalSecondSetter {
-        let mut builder = self.builder;
-        builder.fractional_seconds = Some(Mantissa::Digits(precision_digits));
-        builder.nanoseconds = Some(nanoseconds);
-        FractionalSecondSetter { builder }
-    }
-
-    pub fn with_fractional_seconds(self, fractional_seconds: Decimal) -> FractionalSecondSetter {
-        let mut builder = self.builder;
-        builder.fractional_seconds = Some(Mantissa::Arbitrary(fractional_seconds));
-        builder.nanoseconds = None;
-        FractionalSecondSetter { builder }
+    ) -> TimestampBuilder<HasFractionalSeconds> {
+        self.fractional_seconds = Some(Mantissa::Digits(precision_digits));
+        self.nanoseconds = Some(nanoseconds);
+        self.change_state()
     }
 
-    pub fn build_at_offset(mut self, offset_minutes: i32) -> IonResult<Timestamp> {
-        self.builder.offset = Some(offset_minutes);
-        self.into_builder().build()
+    pub fn with_fractional_seconds(
+        mut self,
+        fractional_seconds: Decimal,
+    ) -> TimestampBuilder<HasFractionalSeconds> {
+        self.fractional_seconds = Some(Mantissa::Arbitrary(fractional_seconds));
+        self.nanoseconds = None;
+        self.change_state()
     }
 
-    /// Like [Self::build_at_offset], but the fields provided for each time unit are understood
-    /// to be in UTC rather than in the local time of the specified offset.
-    pub fn build_utc_fields_at_offset(mut self, offset_minutes: i32) -> IonResult<Timestamp> {
-        self.builder.fields_are_utc = true;
-        self.build_at_offset(offset_minutes)
-    }
-
-    pub fn build_at_unknown_offset(mut self) -> IonResult<Timestamp> {
-        self.builder.offset = None;
-        self.into_builder().build()
-    }
+    with_offset!();
 }
 
-trait TimeUnitSetter {
-    fn into_builder(self) -> TimestampBuilder;
-}
-
-macro_rules! impl_time_unit_setter_for {
-    ($type_name:ty) => {
-        impl TimeUnitSetter for $type_name {
-            fn into_builder(self) -> TimestampBuilder {
-                self.builder
-            }
-        }
-    };
+#[derive(Debug, Clone)]
+pub struct HasFractionalSeconds;
+impl TimestampBuilder<HasFractionalSeconds> {
+    with_offset!();
 }
 
-impl_time_unit_setter_for!(MonthSetter);
-impl_time_unit_setter_for!(DaySetter);
-impl_time_unit_setter_for!(HourAndMinuteSetter);
-impl_time_unit_setter_for!(SecondSetter);
-impl_time_unit_setter_for!(FractionalSecondSetter);
+#[derive(Debug, Clone)]
+pub struct HasOffset;
+// No impl for TimestampBuilder<HasOffset> because `build()` is included in TimestampBuilder<T>
```

Comment:
What does it mean to call `build()` on a `TimestampBuilder<Init>`?

Reply:
You would get `0000T`. 
I could remove that if you think it's too weird.